compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/call/FirQualifiedSupertypeExtendedByOtherSupertypeChecker.kt

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.checkers.call
+
+import org.jetbrains.kotlin.fir.FirSourceElement
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.checkers.isSupertypeOf
+import org.jetbrains.kotlin.fir.analysis.checkers.toRegularClass
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.declarations.FirAnonymousObject
+import org.jetbrains.kotlin.fir.declarations.FirClass
+import org.jetbrains.kotlin.fir.declarations.FirRegularClass
+import org.jetbrains.kotlin.fir.expressions.FirQualifiedAccessExpression
+import org.jetbrains.kotlin.fir.references.FirSuperReference
+import org.jetbrains.kotlin.fir.resolve.transformers.firClassLike
+import org.jetbrains.kotlin.utils.addToStdlib.safeAs
+
+object FirQualifiedSupertypeExtendedByOtherSupertypeChecker : FirQualifiedAccessChecker() {
+    override fun check(functionCall: FirQualifiedAccessExpression, context: CheckerContext, reporter: DiagnosticReporter) {
+        // require to be called over a super reference
+        val superReference = functionCall.calleeReference.safeAs<FirSuperReference>()
+            ?: return
+
+        val explicitType = superReference.superTypeRef.toRegularClass(context.session)
+            ?: return
+
+        val surroundingType = context.findClosestClass()
+            ?: return
+
+        // how many supertypes of the `surroundingType`
+        // have `explicitType` as their supertype
+        var count = 0
+
+        for (it in surroundingType.superTypeRefs) {
+            val that = it.firClassLike(context.session).safeAs<FirClass<*>>()
+                ?: return
+
+            if (
+                explicitType == that ||
+                explicitType.isSupertypeOf(that)
+            ) {
+                count += 1
+
+                if (count >= 2) {
+                    break
+                }
+            }
+        }
+
+        if (count >= 2) {
+            reporter.report(superReference.superTypeRef.source)
+        }
+    }
+
+    /**
+     * Returns the closest to the end of context.containingDeclarations
+     * item like FirRegularClass or FirAnonymousObject
+     * or null if no such item could be found.
+     *
+     * TODO: This function repeats the one in FirNotASupertypeChecker
```

Comment:
You can move this extension to some `ContextUtils.kt` to avoid duplicating code

Reply:
Thanks or the advice I'll definitely follow it. Although I'm not quite certain about the name of the function. I doubt `findClosestClassOrAnonymousObject()` is a good one (btw, there's already the `findClosest<T>()` inside of the context class).

Reply:
I updated the code and put the function `findClosestClassOrObject()` into `FirHelpers.kt` - the file I use to store some helper functions for checkers. It may be worth splitting it into separate ones like `ContextUtils.kt` and other, but I want to do it later.